ta4j-core/src/test/java/org/ta4j/core/ExternalCriterionTest.java

```diff
@@ -0,0 +1,33 @@
+/*
+  The MIT License (MIT)
+
+  Copyright (c) 2014-2017 Marc de Verdelhan & respective authors (see AUTHORS)
+
+  Permission is hereby granted, free of charge, to any person obtaining a copy of
+  this software and associated documentation files (the "Software"), to deal in
+  the Software without restriction, including without limitation the rights to
+  use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+  the Software, and to permit persons to whom the Software is furnished to do so,
+  subject to the following conditions:
+
+  The above copyright notice and this permission notice shall be included in all
+  copies or substantial portions of the Software.
+
+  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+  FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+  COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+  IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+  CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package org.ta4j.core;
+
+public interface ExternalCriterionTest {
+
+    public TimeSeries getSeries() throws Exception;
+
+    public <P> Decimal getFinalCriterionValue(P... params) throws Exception;
```

Comment:
Heap pollution is possible so that the user could get an exception with unreadable stack trace. I don't know if this is important to consider in test package where no changes will be make by user... 

https://stackoverflow.com/questions/12462079/potential-heap-pollution-via-varargs-parameter

Reply:
Thanks!  I'll keep working on this.  How would you feel about defining params to be Decimal varargs instead of generic varargs?  Shouldn't indicator and criteria params be Decimal anyways?

Reply:
Often they are also Integers for the time frame or double for criterions. I dont know what "best practice', maybe we could design an additional class for the parameters like:

```java
 IndicatorParameter{
    private final Type type;
    private final Object value;
   
    public IndicatorParameter(Type type, Object value){
        //...
    }
    
    public getValue(){
        switch(this.type){
            case INTEGER:
                return (int) value;
            //...
        }

    }
    
}
```
or we just handle the initialized indicator with another functional interface that describes how to set the parameters of excel file.

or we just offer functionality to read TimeSeries, Results and parameters from Excel and let the creator of each test class do the mapping and transforming

 
Java doc would be nice, so one can get a faster overview, especially for the excel sheet structure.

Reply:
> public getValue(){

What's the return type?  ;)

A simpler case of the same problem (without thinking about XLS) is simply calling CriterionTest#getCriterion() or IndicatorTest#getIndicator().  How can you call getCriterion() if you have mixed types of parameters?  You wouldn't be able to specify a single signature for the functional interface.  For example, consider two criteria with constructors A#A(int), and B#B(int, double).

My thoughts are now:
1) Use CriterionTest#getCriterion(BigDecimal... params) because BD can be made from int/double/Decimal without rounding and perf in test doesn't matter too much.  Implement ATest#getCriterion(int) that makes a BD[] from the int and then calls CriterionTest#getCriterion(BD[]).  In fact, all criteria could have a BD[] constructor that converts back to int/double, right?
2) Don't use factories b/c of the variability of param signatures and unnecessary complexity.  Implement ATest#getCriterion(int) that calls A#A(int) constructor directly.  This still gives easy copy/paste/edit to BTest and is easy to understand.
3) Implement a Parameters class as you suggested, that can hold different numbers of different types, like with int[] and double[]..  Use CriterionTest#getCriterion(Parameters params).  The use of Parameters in constructors could stay in test or lead to optional constructors in core without removing existing constructors.

I think 3 is best but 2 is easy and simple.  1 is somewhere in between, but would refactor easier to 3.  I'd probably go with 1 unless you feel strongly.  How do you want test and core to look eventually?  Any opinions welcome!

Reply:
> What's the return type? ;)

i forgot:
`public Object getValue()`
The function returns an object that can be casted to the expected value. But take a look at the following, i think we not even need a Parameter class.


We should not use generic varargs in interfaces:

`public <P> Decimal getFinalCriterionValue(P... params) throws Exception;`

Generic varargs will be transformed to array and an array is **non-reifiable** (runtime representation has less information than compile-time representation). So there could be thrown a ClassCastException **at runtime**, which should be avoided in any case.
We also cannot declare the varags as safe with `@SafeVarags` annotation in an interface because we could nether guarantee that the varags  will be modified nor that a reference to them is passed to other code.

To solve this problem i would suggest to use Object varags, because inany case the params have to be casted to the correct type in the lambda expression :

```java
package org.ta4j.core;

@FunctionalInterface
public interface IndicatorFactory<I, D> {
    
    // heap pollution not possible
    Indicator<D> getIndicator(I data, Object... parameters) throws Exception;
}
```